### ISSUE TITLE
Ability to send any TF Build command

### ISSUE THREAD

**Issue by jnm2** [cake-build/cake#2162]
Since some TF Build commands are undocumented, such as `results.publish`, would you be willing to  expose a public method so that Cake users can opt to type their own command strings? That way users can take responsibility to use TF Build commands which Cake does not have because they are new or undocumented.

This is Cake's current implementation, used internally:

https://github.com/cake-build/cake/blob/f11b299d7d4b9782874cb0a9b6c9dee03219ac8b/src/Cake.Common/Build/TFBuild/TFBuildCommands.cs#L321-L328

It's doing the equivalent of vsts-task-lib's [`command`](https://github.com/Microsoft/vsts-task-lib/blob/7dca1fc53e9886ac719a98e811b28339de108de0/node/internal.ts#L277-L280) ([`_writeline`](https://github.com/Microsoft/vsts-task-lib/blob/7dca1fc53e9886ac719a98e811b28339de108de0/node/internal.ts#L52-L54), [`TaskCommand.toString`](https://github.com/Microsoft/vsts-task-lib/blob/7dca1fc53e9886ac719a98e811b28339de108de0/node/taskcommand.ts#L26-L41)), so I'd suggest that we keep the same parameter names and types as the ones which are exported from that public library.

Maybe something like this?

```diff
namespace Cake.Common.Build.TFBuild
{
    public interface ITFBuildCommands
    {
        void WriteWarning(string message);
        void WriteWarning(string message, TFBuildMessageData data);
        void WriteError(string message);
        void WriteError(string message, TFBuildMessageData data);
+       void WriteCommand(string command, IReadOnlyDictionary<string, string> properties, string message);
        void SetProgress(int progress, string currentOperation);
        void CompleteCurrentTask();
        void CompleteCurrentTask(TFBuildTaskResult result);
        Guid CreateNewRecord(string name, string type, int order);
        Guid CreateNewRecord(string name, string type, int order, TFBuildRecordData data);
        void UpdateRecord(Guid id, TFBuildRecordData data);
        void SetVariable(string name, string value);
        void SetSecretVariable(string name, string value);
        void UploadTaskSummary(FilePath markdownPath);
        void UploadTaskLogFile(FilePath logFile);
        void LinkArtifact(string name, TFBuildArtifactType type, string location);
        void UploadArtifact(string folderName, FilePath file);
        void UploadArtifact(string folderName, FilePath file, string artifactName);
        void UploadBuildLogFile(FilePath logFile);
        void UpdateBuildNumber(string buildNumber);
        void AddBuildTag(string tag);
    }
}
```

And it would be used like this ([`TestPublisher.publish`](https://github.com/Microsoft/vsts-task-lib/blob/7dca1fc53e9886ac719a98e811b28339de108de0/node/task.ts#L1684-L1714))?
```cake
TFBuild.Commands.WriteCommand("results.publish", new Dictionary<string, string>
{
    ["type"] = "NUnit",
    ["runTitle"] = "Integration tests",
    ["resultFiles"] = testResultFile.FullPath,
    ["publishRunAttachments"] = "true"
}, message: null);
```

**Comment by gep13** [comment]
@cake-build/cake-team what are your collective thoughts on this one?

**Comment by mholo65** [comment]
Sure, it gives more flexibility. I’ve sometimes wished for same functionality in TeamCity provider also.

**Comment by louisfischer** [comment]
@gep13 and @jnm2 i have similar pull request #2109 for Issue #1609  for this as well to add upload test results and code coverage. The thing is that some of these commands are only support in TFS2017+ or VSTS.

**Comment by jnm2** [comment]
@louisfischer I think the concern Gary expressed in the chat was that committing to an undocumented protocol could result things breaking and it being Cake's responsibility, whereas this PR would allow it to be the user's responsibility to respond to a breaking change in an undocumented protocol. I do like your PR too though. Wouldn't mind having both! 😃

**Comment by gtbuchanan** [comment]
This would be great. I just had to resort to manual logging commands with `Information` to upload artifacts by a directory path since there aren't Cake commands to support it yet. Something like this would be a lot less ugly.

**Comment by agc93** [comment]
@gtbuchanan What were you looking to do that wasn't supported? We've already got the `UploadArtifact` aliases for uploading files (and creating artifacts) that you can see in the API [here](https://cakebuild.net/api/Cake.Common.Build.TFBuild/ITFBuildCommands/).

Just wanting to understand where our gaps are :)

**Comment by gtbuchanan** [comment]
@agc93 `UploadArtifact` in Cake currently only accepts `FilePath` but Azure Pipelines also supports directories. See:

https://github.com/Microsoft/azure-pipelines-tasks/blob/master/Tasks/PublishBuildArtifactsV1/publishbuildartifacts.ts#L50

https://github.com/gtbuchanan/repo-template-cs/blob/master/build.cake#L200

**Comment by patriksvensson** [comment]
@gtbuchanan I think that mandates a separate issue.

**Comment by gtbuchanan** [comment]
@patriksvensson I agree. I was simply responding to the question. Haven't had the time to create another issue/PR yet. Nonetheless, my main point was this feature would be helpful in the interim while waiting for any updates in the future.

**Comment by patriksvensson** [comment]
@gtbuchanan Absolutely. I just don't want a required change to get lost in the comments 😄
